### PR TITLE
feat(core): add architecture/reserved-directory-names

### PR DESCRIPTION
**Pull request by oekazuma**
Adds `architecture/reserved-directory-names`, the charter's M3 mechanism: a directory's immediate subdirectories may only take names the project declared for that position. Like the other Architecture convention rules it is L3 — it emits nothing until a scope is declared.

```js
'architecture/reserved-directory-names': {
  options: {
    scopes: { 'src/lib': 'api|components|features|effect|db' },
    unitScopes: { 'src/**': 'parts|functions|stores|types|tests|styleGuide' }
  }
}
```

## The two options differ in what their key names

**A `scopes` key names the parent directly.** `'src/lib'` matches `src/lib`, and its immediate subdirectories may take the names listed.

**A `unitScopes` key names a root**, and the rule governs the children of whichever directories beneath it are **units** — the shape a glob cannot reach, because units nest to arbitrary depth. A `scopes` key is only worth writing where the children are entirely drawn from the names listed; a route directory holds its reserved names beside its route segments, and route segments are unbounded, so no declaration belongs there.

When both maps match one directory, specificity decides — more path segments, then fewer `**` segments, then the longer key, then alphabetically first — so either map can narrow the other. Two byte-identical globs are the only pair those steps cannot separate, and there `scopes` wins, because it applies to every directory its key matches while `unitScopes` applies only to the ones that are units. That pair is reported: wherever both are declared, the `unitScopes` entry does nothing.

## What this catches that the two shipped rules cannot

`architecture/directory-naming` checks a directory's **casing**; this checks its **name**. A `helpers/` inside a component unit is perfectly camelCase, so no casing declaration objects to it. Only a closed vocabulary can say that `helpers` is not one of the names this location admits.

## A third definition of "unit", and why

`architecture/unit-entry-file` identifies a candidate unit by its first character and then reports whether the entry file is there. **This rule cannot borrow that definition.** A capitalised directory with no file named after it is a grouping wearing the wrong name — which that rule already reports, once. Treating it as a unit here would measure each of its children against the unit vocabulary, so a directory of capitalised components would produce a finding per component: one naming mistake becoming N findings, none of them naming the real problem.

So a unit here is a directory whose name begins A–Z **and** one of whose immediate children is a file whose name up to the first dot equals the directory's name. Extension-agnostic, so `.svelte`, `.ts` and `.svelte.ts` all qualify, and independent of the sibling rule's configuration. A test pins the non-cascade, because the cheap implementation is the one that cascades.

## Declarations that do not check what they say

One folded finding, as both sibling rules emit, naming which of five reasons applies: the glob matched no directory; every directory it matched is excluded; it matched directories but never a unit; the value lists no name at all; the same glob is declared in both maps. The reasons exist because the remedies differ.

The third is the same distinction the sibling rule draws for its casing-gated map, where the gate *is* the identification criterion — that rule's implementation got it wrong twice, in opposite directions, before it settled, so this one carries it from the design rather than discovering it.

## Shared machinery

The module the two shipped directory rules share gains five primitives — a parent-to-children map, a directory-to-files map, a declaration-value splitter, a glob comparator usable outside the matcher, and a key-matching primitive. Two existing functions are reimplemented on them so the logic exists once, proven behaviour-preserving by their existing tests going untouched.

One premise all three rules rest on is now written down where it belongs: `Runtime.glob` returns files only and excludes dot files, so a directory holding no file at any depth does not exist as far as these rules are concerned and a dot directory never appears. Nothing stated it before, and one of these rules enumerates a parent's children exhaustively.

## Verification

lint, format and typecheck clean; **2,040 tests** pass (core 1048, cli 769, vite 198, mcp 25). The documented example is held to being real by a test asserting it reports nothing on a conforming tree and leaves no declaration reported — silence is proof here rather than absence of proof, because this rule reports a declaration that identified nothing. The `exclude` example is asserted in both directions, and the wiring from both the CLI and the vite plugin is verified to fail if either entry point stops passing the inventory.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added the `architecture/reserved-directory-names` rule to flag directories whose immediate children aren’t in the configured allowlist.
  * Supports `scopes`, `unitScopes`, and `exclude`, and remains off until configured.

* **Documentation**
  * Added English and Japanese docs for the new rule, including configuration options, precedence, and edge cases.
  * Updated Architecture rule listings and configuration rule-option tables.
  * Clarified PascalCase vs related directory naming checks.

* **Tests**
  * Added unit, CLI, and Vite integration coverage for rule behavior, exclusions, and precedence.

### CODE REVIEW

File: docs/src/content/docs/rules/architecture/reserved-directory-names.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+`Card/Card.test.ts` — counts as a unit too, and its children are checked. The alternative, stripping a
+single extension, would reject a real entry-file shape, and a finding a reader can dismiss is the
+milder failure.
+
+The cut applies only to the filename, not the directory name — the directory's basename is compared
+whole. `src/lib/Card.v2/Card.v2.svelte` is accepted: the stem of `Card.v2.svelte` up to its first dot
+is `Card`, which does not equal the directory's own (uncut) name `Card.v2`, so this rule does not
+consider it a unit. `architecture/unit-entry-file`, configured with an explicit extension, asks a
+different question — whether `Card.v2` + `.svelte` exists — and answers yes for the same directory.
+Both answers are consistent with each rule's own definition.
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply docs/src/content/docs/rules/architecture/reserved-directory-names.md:124]: Fixed in 064ac3c0. "Accepted" was meant as "this trade-off is accepted" and read as "the directory is accepted as a unit" — the opposite. Both pages now lead with the outcome (`src/lib/Card.v2/` is **not a unit here**) and add the consequence the old wording left out: its children go unchecked.

File: packages/core/src/rules/architecture/reserved-directory-names.ts [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+      for (const key of Object.keys(scopesMap)) {
+        if (!Object.hasOwn(unitMap, key)) continue;
+        // A `scopes` value naming nothing is dropped before matching, so the `unitScopes` entry does
+        // govern and the claim below would be false. Skipping here also lets the empty-value reason
+        // report the real error, which it cannot do for a key that already carries a note.
+        if (namesOf(scopesMap[key] as string).length === 0) continue;
+        collisions.add(key);
+      }
+    };
+    noteCollisions(globalScopes, globalUnits);
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply packages/core/src/rules/architecture/reserved-directory-names.ts:136]: Fixed in 064ac3c0, both halves. This is the mirror image of a bug the final pre-merge review caught in the other direction, and the asymmetry is exactly what let it survive.

Both guards are needed, and your point about them being needed *together* is the sharp part. Guarding the collision alone would have been worse than leaving it: the `scopes` side governs elsewhere, so `usedKeys` absorbs the key, the unused classification never sees it, and the declaration would end up with no note at all rather than merely the wrong one. Reading both maps in the empty-value loop is what keeps the diagnosis.

Added the mirror test, and confirmed by mutation that the new `unitMap` guard is load-bearing: dropping it makes only that test fail, with the collision wording instead of the empty-value one.

One note on the second diff: I kept `Object.hasOwn` rather than `!== undefined` for both presence checks. A `string-map` value is looked up on a plain object literal here, so a key named `toString` would otherwise resolve through the prototype chain — the same hazard `parseCasings` in `casing.ts` guards against, for the same reason.
